### PR TITLE
Provide a simpler way to transform the objects of `HttpRequest` and `HttpResponse`

### LINKED ISSUE #3606: Provide a way to transform an `HttpHeaders` and `HttpData` in `Http{Request,Response}`
Currently, we proivde `FilteredHttpRequest` and `FilteredHttpResponse` for tranforming a headers or data.
But it is a bit difficult to implement `FilteredHttp*` because users have to check the type of `HttpObject`.
Most users might not know the hierarchical relation between `HttpOject`, `HttpHeaders`, `ResponseHeaders` and `HttpData` 
```java
return new FilteredHttpResponse(response) {
        protected HttpObject filter(HttpObject obj) {
            if (obj instanceof ResponseHeaders) {
               ...
            } else if (obj instanceof HttpData) {
                ...
            } else {
                // Http trailers.
                assert obj instanceof HttpHeaders;
                ...
            }
            return obj;
        }
```

It would be convenient to provide dedicated transforming functions for `HttpRequest` and `HttpResponse`. 

- `Http{Request,Response}.mapHeaders(Function<? super ResponseHeaders, ? extends ResponseHeaders> f)` 
- `Http{Request,Response}.mapData(Function<? super HttpData, ? extends HttpData> f)`
- `Http{Request,Response}.mapTrailers(Function<? super HttpHeaders, ? extends HttpHeaders> f)`

```java
HttpResponse response = ...;
response.mapHeaders(headers -> {
    if (headers.isInformational()) {
        return ob;
    } else {
        return headers.toBuilder()...build();
    }
}).mapData(data -> {
    ...
});
```

I think `StreamMesage.filter(Predicate)` and `StreamMessage.map(Function)` could be used for implementing those functions.

### PR DESCRIPTION
**Pull request by ikhoon**
Motivation:

Currently, we provide FilteredHttpRequest and FilteredHttpResponse for
transforming a headers or data.  It is a bit difficult to implement
`FilteredHttp*` by users because they have to check the type of HttpObject.
Most users might not know the hierarchical relation between `HttpObject`,
`HttpHeaders`, `RequestHeaders`, `ResponseHeaders` and `HttpData`.

See #3606 for details

Modifications:

- Add `Http{Request,Response}.mapHeaders(Function<? super ResponseHeaders, ? extends ResponseHeaders> f)`
  for transforming an headers.
- Add `Http{Request,Response}.mapData(Function<? super HttpData, ? extends HttpData> f)`
  for transforming `HttpData`s
- Add `Http{Request,Response}.mapTrailers(Function<? super HttpHeaders, ? extends HttpHeaders> f)`
  for transforming a trailers.
- Add `HttpResponse.mapInformational(Function<? super ResponseHeaders, ? extends ResponseHeaders> f))`
  for transforming a non-informational headers.

Result:

- You can now transform the objects of `HttpRequest` and `HttpResponse` using a function.
  ```java
  HttpRequest req1 = ...;
  HttpRequest req2 = req1.mapHeaders(headers -> ...)
                         .mapData(data -> ...)
                         .mapTrailers(trailers -> ...);

  HttpResponse res1 = ...;
  HttpResponse res2 = res2.mapHeaders(headres -> ...)
                          .mapInformational(informational -> ...)
                          .mapData(data -> ...)
                          .mapTrailers(trailers -> ...);
  ```
- Fixes #3606

### CODE REVIEW

File: core/src/main/java/com/linecorp/armeria/common/HttpRequest.java [hunk 1]
```diff
@@ -519,4 +523,54 @@ default HttpRequestDuplicator toDuplicator(EventExecutor executor, long maxReque
...
+
+    /**
+     * Transforms the {@link ResponseHeaders} of this {@link HttpRequest} by applying the specified
+     * {@link Function}.
+     */
+    default HttpRequest mapHeaders(Function<? super RequestHeaders, ? extends RequestHeaders> function) {
+        requireNonNull(function, "function");
+        final RequestHeaders transformed = function.apply(headers());
+        requireNonNull(transformed, "function.apply() returned null");
+        return of(transformed, this);
```
Comment (minwoox) [inline core/src/main/java/com/linecorp/armeria/common/HttpRequest.java:535]:
```suggestion
        return withHeaders(transformed);
```

File: core/src/main/java/com/linecorp/armeria/common/HttpRequest.java [hunk 2]
```diff
@@ -519,4 +523,54 @@ default HttpRequestDuplicator toDuplicator(EventExecutor executor, long maxReque
...
+            return obj;
+        });
+        return of(headers(), stream);
+    }
+
+    /**
+     * Transforms the {@linkplain HttpObject}s emitted by this {@link HttpRequest} by applying the
+     * specified {@link Function}.
+     */
+    default HttpRequest mapObject(Function<? super HttpObject, ? extends HttpObject> function) {
```
Comment (minwoox) [inline core/src/main/java/com/linecorp/armeria/common/HttpRequest.java:568]: How about adding this method when we need it?

File: core/src/main/java/com/linecorp/armeria/common/HttpRequest.java [hunk 3]
```diff
@@ -519,4 +523,54 @@ default HttpRequestDuplicator toDuplicator(EventExecutor executor, long maxReque
...
+    }
+
+    /**
+     * Transforms the {@link HttpData}s emitted by this {@link HttpRequest} by applying the specified
+     * {@link Function}.
+     */
+    default HttpRequest mapData(Function<? super HttpData, ? extends HttpData> function) {
+        requireNonNull(function, "function");
+        final StreamMessage<HttpObject> stream =
+                map(obj -> obj instanceof HttpData ? function.apply((HttpData) obj) : obj);
```
Comment (minwoox) [inline core/src/main/java/com/linecorp/armeria/common/HttpRequest.java:545]: I think this does not fuse when multiple
`request.mapData(...).mapData(...).mapTrailers(...)` is called which means it creates additional instance every time.
How about checking if the delegate of `StreamMessageBasedHttpRequest` is fuseable and fusing it ?

File: core/src/main/java/com/linecorp/armeria/common/HttpResponse.java [hunk 4]
```diff
@@ -559,4 +562,50 @@ default SplitHttpResponse split(EventExecutor executor) {
...
     }
+
+    /**
+     * Transforms the {@link ResponseHeaders} emitted by {@link HttpResponse} by applying the specified
+     * {@link Function}.
+     */
+    default HttpResponse mapHeaders(Function<? super ResponseHeaders, ? extends ResponseHeaders> function) {
+        requireNonNull(function, "function");
+        final StreamMessage<HttpObject> stream =
+                map(obj -> obj instanceof ResponseHeaders ? function.apply((ResponseHeaders) obj) : obj);
```
Comment (minwoox) [inline core/src/main/java/com/linecorp/armeria/common/HttpResponse.java:573]: ditto

File: core/src/main/java/com/linecorp/armeria/common/HttpResponse.java [hunk 5]
```diff
@@ -559,4 +562,50 @@ default SplitHttpResponse split(EventExecutor executor) {
...
+            return obj;
+        });
+        return of(stream);
+    }
+
+    /**
+     * Transforms the {@linkplain HttpObject}s emitted by this {@link HttpRequest} by applying the
+     * specified {@link Function}.
+     */
+    default HttpResponse mapObject(Function<? super HttpObject, ? extends HttpObject> function) {
```
Comment (minwoox) [inline core/src/main/java/com/linecorp/armeria/common/HttpResponse.java:607]: ditto. How about adding it when we need it?

**Review by minwoox — COMMENTED** [review]
Left some comments.
It's much cleaner now. 👍

**Comment by minwoox** [comment]
https://github.com/line/armeria/pull/3624/checks?check_run_id=2808614800#step:7:1951
Oops, we have a test that failed. 😄

**Comment by ikhoon** [comment]
> [#3624 (checks)](https://github.com/line/armeria/pull/3624/checks?check_run_id=2808614800#step:7:1951)
> Oops, we have a test that failed. 😄

😱

File: core/src/main/java/com/linecorp/armeria/common/HttpResponse.java [hunk 6]
```diff
@@ -569,4 +572,41 @@ default SplitHttpResponse split(EventExecutor executor) {
...
     }
+
+    /**
+     * Transforms the {@link ResponseHeaders} emitted by {@link HttpResponse} by applying the specified
+     * {@link Function}.
+     */
+    default HttpResponse mapHeaders(Function<? super ResponseHeaders, ? extends ResponseHeaders> function) {
+        requireNonNull(function, "function");
+        final StreamMessage<HttpObject> stream =
+                map(obj -> obj instanceof ResponseHeaders ? function.apply((ResponseHeaders) obj) : obj);
```
Comment (minwoox) [inline core/src/main/java/com/linecorp/armeria/common/HttpResponse.java:583]: How about applying the `function` only when the `ResponseHeaders` is not informational?
I think there's no chances that a user want to manipulate the information headers.
Also users do not have to check if it's informational in the `function`.

File: core/src/main/java/com/linecorp/armeria/common/HttpRequest.java [hunk 7]
```diff
@@ -525,4 +528,41 @@ default HttpRequestDuplicator toDuplicator(EventExecutor executor, long maxReque
...
+    }
+
+    /**
+     * Transforms the {@linkplain HttpHeaders trailers} emitted by this {@link HttpRequest} by applying the
+     * specified {@link Function}.
+     */
+    default HttpRequest mapTrailers(Function<? super HttpHeaders, ? extends HttpHeaders> function) {
+        requireNonNull(function, "function");
+        final StreamMessage<HttpObject> stream = map(obj -> {
+            if (obj instanceof HttpHeaders && !(obj instanceof RequestHeaders)) {
```
Comment (minwoox) [inline core/src/main/java/com/linecorp/armeria/common/HttpRequest.java:561]: We probably don't need `!(obj instanceof RequestHeaders)`?

**Review by minwoox — COMMENTED** [review]
Thanks! Left a couple of nits. 😄

File: core/src/main/java/com/linecorp/armeria/common/HttpResponse.java (hunk 6 above)
Comment (ikhoon) [reply core/src/main/java/com/linecorp/armeria/common/HttpResponse.java:583]: I am not sure 100% that it is a good decision to exclude an informational headers from `mapHeaders()` without additional options.
How about adding `mapInformational(Function)` for an informational headers?

Comment (minwoox) [reply core/src/main/java/com/linecorp/armeria/common/HttpResponse.java:583]: That also sounds good. 👍

File: core/src/main/java/com/linecorp/armeria/common/HttpRequest.java (hunk 7 above)
Comment (ikhoon) [reply core/src/main/java/com/linecorp/armeria/common/HttpRequest.java:561]: Ah... nice advice!

**Review by minwoox — APPROVED** [review]
Nice addition.
Thanks! 👍

**Review by trustin — APPROVED** [review]

**Pull request merged**